File: VB/BackgroundGeolocation.Module/BusinessObjects/DeviceInfo.vb
Imports System
Imports DevExpress.ExpressApp.Model
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace BackgroundGeolocation.Module.BusinessObjects
	<DefaultClassOptions>
	Public Class DeviceInfo
		Inherits BaseObject

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

        Private fName As String
        Public Property Name() As String
            Get
                Return fName
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Name", fName, value)
            End Set
		End Property

        Private fUuid As String
        Public Property Uuid() As String
            Get
                Return fUuid
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Uuid", fUuid, value)
            End Set
		End Property

        Private fModel As String
        Public Property Model() As String
            Get
                Return fModel
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Model", fModel, value)
            End Set
		End Property

        Private fPlatform As String
        Public Property Platform() As String
            Get
                Return fPlatform
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Platform", fPlatform, value)
            End Set
		End Property

        Private fVersion As String
        Public Property Version() As String
            Get
                Return fVersion
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Version", fVersion, value)
            End Set
		End Property

        Private fManufacturer As String
        Public Property Manufacturer() As String
            Get
                Return fManufacturer
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Manufacturer", fManufacturer, value)
            End Set
		End Property

		<Association("DeviceLocation-Locations")>
		Public ReadOnly Property Locations() As XPCollection(Of DeviceLocation)
			Get
				Return GetCollection(Of DeviceLocation)("Locations")
			End Get
		End Property
	End Class

	'[DefaultClassOptions]
	Public Class DeviceLocation
		Inherits BaseObject
		Implements IMapsMarker

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

        Private fDeviceInfo As DeviceInfo
        <Association("DeviceLocation-Locations")>
        Public Property DeviceInfo() As DeviceInfo
            Get
                Return fDeviceInfo
            End Get
            Set(ByVal value As DeviceInfo)
                SetPropertyValue("DeviceInfo", fDeviceInfo, value)
            End Set
		End Property

        Private fTimeStamp As Date
        <ModelDefault("DisplayFormat", "{0: dd/MM/yyyy HH:mm:ss}")>
        Public Property TimeStamp() As Date
            Get
                Return fTimeStamp
            End Get
            Set(ByVal value As Date)
                SetPropertyValue("TimeStamp", fTimeStamp, value)
            End Set
		End Property

		Public ReadOnly Property Title() As String Implements DevExpress.Persistent.Base.IBaseMapsMarker.Title
			Get
				Return String.Format("Time: {0}<br>Accuracy: {1}<br>Speed: {2}<br>Activity: {3}", TimeStamp, Accuracy, Speed, ActivityType)
			End Get
		End Property

        Private fLatitude As Double
        Private ReadOnly Property IBaseMapsMarker_Latitude() As Double Implements DevExpress.Persistent.Base.IBaseMapsMarker.Latitude
            Get
                Return Latitude
            End Get
        End Property
        Public Property Latitude() As Double
            Get
                Return fLatitude
            End Get
            Set(ByVal value As Double)
                SetPropertyValue("Latitude", fLatitude, value)
            End Set
		End Property

        Private fLongitude As Double
        Private ReadOnly Property IBaseMapsMarker_Longitude() As Double Implements DevExpress.Persistent.Base.IBaseMapsMarker.Longitude
            Get
                Return Longitude
            End Get
        End Property
        Public Property Longitude() As Double
            Get
                Return fLongitude
            End Get
            Set(ByVal value As Double)
                SetPropertyValue("Longitude", fLongitude, value)
            End Set
		End Property

        Private fAccuracy As Double
        Public Property Accuracy() As Double
            Get
                Return fAccuracy
            End Get
            Set(ByVal value As Double)
                SetPropertyValue("Accuracy", fAccuracy, value)
            End Set
		End Property

        Private fSpeed As Double
        Public Property Speed() As Double
            Get
                Return fSpeed
            End Get
            Set(ByVal value As Double)
                SetPropertyValue("Speed", fSpeed, value)
            End Set
		End Property

        Private fActivityType As String
        Public Property ActivityType() As String
            Get
                Return fActivityType
            End Get
            Set(ByVal value As String)
                SetPropertyValue("ActivityType", fActivityType, value)
            End Set
		End Property
	End Class
End Namespace

File: CS/BackgroundGeolocation.Module/BusinessObjects/DeviceInfo.cs
using System;
using DevExpress.ExpressApp.Model;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;

namespace BackgroundGeolocation.Module.BusinessObjects {
    [DefaultClassOptions]
    public class DeviceInfo : BaseObject {
        public DeviceInfo(Session session) : base(session) { }

        private string name;
        public string Name {
            get { return name; }
            set { SetPropertyValue("Name", ref name, value); }
        }

        private string uuid;
        public string Uuid {
            get { return uuid; }
            set { SetPropertyValue("Uuid", ref uuid, value); }
        }

        private string model;
        public string Model {
            get { return model; }
            set { SetPropertyValue("Model", ref model, value); }
        }

        private string platform;
        public string Platform {
            get { return platform; }
            set { SetPropertyValue("Platform", ref platform, value); }
        }

        private string version;
        public string Version {
            get { return version; }
            set { SetPropertyValue("Version", ref version, value); }
        }

        private string manufacturer;
        public string Manufacturer {
            get { return manufacturer; }
            set { SetPropertyValue("Manufacturer", ref manufacturer, value); }
        }

        [Association("DeviceLocation-Locations")]
        public XPCollection<DeviceLocation> Locations {
            get { return GetCollection<DeviceLocation>("Locations"); }
        }
    }

    //[DefaultClassOptions]
    public class DeviceLocation : BaseObject, IMapsMarker {
        public DeviceLocation(Session session) : base(session) { }

        private DeviceInfo deviceInfo;
        [Association("DeviceLocation-Locations")]
        public DeviceInfo DeviceInfo {
            get { return deviceInfo; }
            set { SetPropertyValue("DeviceInfo", ref deviceInfo, value); }
        }

        private DateTime timeStamp;
        [ModelDefault("DisplayFormat", "{0: dd/MM/yyyy HH:mm:ss}")]
        public DateTime TimeStamp {
            get { return timeStamp; }
            set { SetPropertyValue("TimeStamp", ref timeStamp, value); }
        }

        public string Title {
            get {
                return String.Format("Time: {0}<br>Accuracy: {1}<br>Speed: {2}<br>Activity: {3}", TimeStamp, Accuracy, Speed, ActivityType);
            }
        }

        private double latitude;
        public double Latitude {
            get { return latitude; }
            set { SetPropertyValue("Latitude", ref latitude, value); }
        }

        private double longitude;
        public double Longitude {
            get { return longitude; }
            set { SetPropertyValue("Longitude", ref longitude, value); }
        }

        private double accuracy;
        public double Accuracy {
            get { return accuracy; }
            set { SetPropertyValue("Accuracy", ref accuracy, value); }
        }

        private double speed;
        public double Speed {
            get { return speed; }
            set { SetPropertyValue("Speed", ref speed, value); }
        }

        private string activityType;
        public string ActivityType {
            get { return activityType; }
            set { SetPropertyValue("ActivityType", ref activityType, value); }
        }
    }
}
